error.go

```diff
@@ -371,21 +370,77 @@ func (e errParamGroupFailed) updateGraph(g *dot.Graph) {
 	g.FailGroupNodes(e.Key.group, e.Key.t, e.CtorID)
 }
 
-// errMissingType is returned when a single value that was expected in the
-// container was not available.
-type errMissingType struct {
-	Key key
+// missingType holds information about a type that was missing in the
+// container.
+type missingType struct {
+	Key key // item that was missing
 
 	// If non-empty, we will include suggestions for what the user may have
 	// meant.
 	suggestions []key
 }
 
-func newErrMissingType(c containerStore, k key) errMissingType {
+// Format prints a string representation of missingType.
+//
+// With %v, it prints a short representation ideal for an itemized list.
+//
+//   io.Writer
+//   io.Writer: did you mean *bytes.Buffer?
+//   io.Writer: did you mean *bytes.Buffer, or *os.File?
+//
+// With %+v, it prints a longer representation ideal for standalone output.
```

Comment:
This point was initially counter-intuitive to me. But the point you are trying to make is that in the world of "+v" or multiline the printed comments can be more verbose because you have more space.

Why not always opt for the shorter message? Even in the shortest form, the message has to be well understood for a minimum UX. If the message is already well understood, why not just use the short message always? The net benefit is less code to maintain.

Reply:
This divergence was made with the intent of being close to the existing output. Before errMissingType and errManyMissingTypes were merged into errMissingTypes, we had that distinction in the output. The idea was to reduce the number of tests that had to be updated in this change, since we can iterate on the error messages later as well.

Reply:
Okay, understood. But do you agree that eventually there should only be one error message?

Reply:
Yeah, they should be similar. If we were writing this error type from scratch, we wouldn't have the two varying error messages.